### PR TITLE
[ML] Improvement: releasing unused memory and unnecessary copy of buffers in ScrollDataExtractor

### PR DESCRIPTION
**Pull request by parasmalik**
All tests of the class are fixed and working fine. Task :server:test is failing even for elasticsearch main repo( without my changes). I might be some environment issue of my laptop.

### CODE REVIEW

File: docs/changelog/98324.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+pr: 98324
+summary: Releasing unused memory and unnecessary copy of buffers in ScrollDataExtractor
+area: ML
+type: improvement
```
Comment (droberts195) [inline docs/changelog/98324.yaml:4]:
```suggestion
summary: Avoid risk of OOM in datafeeds when memory is constrained
area: Machine Learning
type: bug
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java [hunk 2]
```diff
@@ -166,18 +169,37 @@ private SearchRequestBuilder buildSearchRequest(long start) {
         return searchRequestBuilder;
     }
 
-    private InputStream processSearchResponse(SearchResponse searchResponse) throws IOException {
+    /*
+        Utility class to convert ByteArrayOutputStream to ByteArrayInputStream without copying the underlying buffer.
+     */
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java:174]:
```suggestion
    /**
     * Utility class to convert ByteArrayOutputStream to ByteArrayInputStream without copying the underlying buffer.
     */
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java [hunk 3]
```diff
@@ -166,18 +169,37 @@ private SearchRequestBuilder buildSearchRequest(long start) {
...
+    private class ConvertableByteArrayOutputStream extends ByteArrayOutputStream {
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java:175]:
```suggestion
    private static class ConvertableByteArrayOutputStream extends ByteArrayOutputStream {
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java [hunk 4]
```diff
@@ -166,18 +169,37 @@ private SearchRequestBuilder buildSearchRequest(long start) {
...
+            return inputStream;
+        }
+    }
 
-        scrollId = searchResponse.getScrollId();
-        if (searchResponse.getHits().getHits().length == 0) {
+    /*
+    IMPORTANT: This is not a idempotent method. This method change the input parameter SearchResponse.
+    It deletes are all SearchHits from input object SearchResponse.
+     */
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java:187]:
```suggestion
    /**
     * IMPORTANT: This is not an idempotent method. This method changes the input array by setting each element to <code>null</code>.
     */
```

**Review by droberts195 — COMMENTED** [review]
Thanks for the contribution @i-plusplus.

I left a few initial suggestions. Let's also see if the CI run picks up any problems as a result.

**Comment by droberts195** [comment]
@elasticmachine ok to test

**Comment by i-plusplus** [comment]
Thanks @droberts195. I have merged your suggestions and all Checks are passed :).

**Review by droberts195 — APPROVED** [review]
LGTM

Thanks again for the contribution @i-plusplus.

I will merge this now.

**Pull request merged**